packages/fuels-programs/src/lib.rs

```diff
@@ -21,3 +32,323 @@ impl Configurables {
+        let mut receipts = vec![
+            Receipt::Call {
+                id: Default::default(),
+                to: Default::default(),
+                amount: Default::default(),
+                asset_id: Default::default(),
+                gas: Default::default(),
+                param1: Default::default(),
+                param2: Default::default(),
+                pc: Default::default(),
+                is: Default::default(),
+            },
+            Receipt::Revert {
+                id: Default::default(),
+                ra: Default::default(),
+                pc: Default::default(),
+                is: Default::default(),
+            },
+            Receipt::Log {
+                id: Default::default(),
+                ra: Default::default(),
+                rb: Default::default(),
+                rc: Default::default(),
+                rd: Default::default(),
+                pc: Default::default(),
+                is: Default::default(),
+            },
+            Receipt::LogData {
+                id: Default::default(),
+                ra: Default::default(),
+                rb: Default::default(),
+                ptr: Default::default(),
+                len: Default::default(),
+                digest: Default::default(),
+                data: Default::default(),
+                pc: Default::default(),
+                is: Default::default(),
+            },
+            Receipt::ScriptResult {
+                result: ScriptExecutionResult::Success,
+                gas_used: Default::default(),
+            },
+        ];
+        let relevant_receipts = get_relevant_receipts();
+        receipts.extend(relevant_receipts.clone());
+
+        let parser = ReceiptParser::new(&receipts);
+
+        assert_eq!(parser.receipts, relevant_receipts);
+
+        Ok(())
+    }
+
+    #[tokio::test]
+    async fn receipt_parser_empty_receipts() -> Result<()> {
+        let receipts = [];
+        let mut parser = ReceiptParser::new(&receipts);
+        let output_param = ParamType::Unit;
+
+        let error = parser
+            .try_parse_output(Default::default(), &output_param)
+            .expect_err("should error");
+
+        let expected_error = ReceiptParser::parsing_error(&output_param);
+        assert_eq!(error.to_string(), expected_error.to_string());
+
+        Ok(())
+    }
+
+    #[tokio::test]
+    async fn receipt_parser_extract_return_data() -> Result<()> {
+        let expected_receipts = get_relevant_receipts();
+
+        let mut receipts = expected_receipts.clone();
+        receipts.push(get_return_data_receipt(target_contract(), ENCODED_DATA));
+        let mut parser = ReceiptParser::new(&receipts);
+
+        let encoded_data = parser
+            .extract_return_data(target_contract())
+            .expect("This should return data");
+
+        assert_eq!(encoded_data, ENCODED_DATA);
+        assert_eq!(parser.receipts, expected_receipts);
+
+        Ok(())
```

Comment:
It is better to test through the public interface whenever we can -- makes the component more friendly to refactorings since you won't be battling unit tests that refer to your private innards all the time.

Also inspires more confidence in the component since you're using it in the tests the same way other parts of the code base will use it.
```suggestion
        let expected_receipts = get_relevant_receipts();
        let contract_id = target_contract();

        let mut receipts = expected_receipts.clone();
        receipts.push(get_return_receipt(contract_id, ENCODED_VAL));
        let mut parser = ReceiptParser::new(&receipts);

        let token = parser
            .parse(Some(&contract_id.into()), &u64::param_type())
            .expect("parsing should succeed");

        assert_eq!(u64::from_token(token)?, 225u64);
        assert_eq!(parser.receipts, expected_receipts);

        Ok(())
```

Reply:
The goal here was to test the private innards directly, ensuring that small changes to them don't lead to unwanted consequences. These tests can be removed should there be further reaching changes to how we approach receipt parsing.
The correct behaviour of the parser as a whole is tested on a higher level already by our e2e tests.

Reply:
> The goal here was to test the private innards directly, ensuring that small changes to them don't lead to unwanted consequences.

I understand, but private innards are implementation details. Caring exactly about how they work damages the refactorability of the code. If you had another idea of how to structure the innards of the receipt parser, the tests are going to fight you rather than help.

A common practice for other languages is to have the unit tests reside in another file. As such, they cannot access the private innards of the thing they're testing. So that forces you into a choice, either test through the public interface or redesign. The reasoning is that if you cannot test through the public interface then you've made your component too big and should consider delegating the work across multiple components.

I'm not innocent of doing this myself when it was a hassle to test through the public interface. But here it requires a little extra effort.

> These tests can be removed should there be further reaching changes to how we approach receipt parsing.

Yes, but by removing them you remove the protection they offer. So you have to write new ones when refactoring. This is the damage to refactorability that I'm referring to.

> The correct behaviour of the parser as a whole is tested on a higher level already by our e2e tests.

I agree that it would be hard to have all the e2e tests pass and not have this component work properly. 

The point I want to get across is that unit tests should completely test the component in question. This offers most of the protection of the e2e tests at a fraction of the cost to run. e2e is to be reserved for validating our expectations against the node or a few select happy paths to see everything integrated correctly. We shouldn't rely on e2e tests to cover what we could have caught in unit tests but didn't. The testing pyramid is a pyramid for a reason. Continuing down the path of needless delegation to e2e testing will have us wait an hour for the suite to complete before long.

Reply:
Makes sense, thanks for the thorough feedback ❤️